### PR TITLE
Child/dependent tax credit

### PR DESCRIPTION
**Pull request by NathanielLovin**
Attempting to start work on implementing the Child Tax Credit and Other Dependent Tax Credit (line 19).

To begin with, I implemented the two checkboxes from #70. The next step is to go through the worksheet for the credit and start implementing the calculations.

I also made it so the dependent form closes when the dependent is added, instead of remaining open to add more dependents. I think it's more user friendly, but remove it if you want, it's just the `newDependent(false)` line in `onSubmit`.

### CODE REVIEW

**Comment by NathanielLovin** [comment]
For a part of the Additional Child Tax Credit, it asks for Social Security and Medicare withholdings from W2. I added those fields to the W-2 input field as required, but we could make them optional.

I did not add complete coverage for either line 1 or line 6 of Schedule 8812 because there are a lot of edge cases that aren't in the system yet and are more difficult to pre-code like we're doing with the schedules. But it's mostly there. I attempted to document the decisions of things to exclude as well as possible.

Not ready for full review (needs more testing on my end), but any thoughts would be appreciated.

**Comment by zakpatterson** [comment]
You're being quite comprehensive in your approach to this, which is admirable. I think it's acceptable to get the credit working for a decent subset of valid cases, and leave todos for edge cases. Sounds like you're doing that, which is great.

I spotted this comment and lol'd a little. I worked on some of those EIC cases today: https://github.com/ustaxes/UsTaxes/pull/191/files#diff-804a3078513157524b5d542c203db59ba5dad140a5ac10ea89d5b69eb2fd2563R26

So yes, the EIC credit has a bunch of rules about what is considered earned income. At this point I'm leaving those items as documented but not implemented, and I think they should also be logger warnings.

**Comment by zakpatterson** [comment]
I'm attempting to assist with merge conflicts and finishing credit for other dependents, here: https://github.com/NathanielLovin/UsTaxes/pull/1

**Comment by NathanielLovin** [comment]
I've added the Schedule 8812 attachment and also made some other corrections to the code, along with adding the EIC earned income definition if EIC is being filled.

I believe this is ready to be reviewed now.

**Comment by zakpatterson** [comment]
I took a look at this in terms of behavior and it looks great. Can you display the child tax credit status on the Summary page?

File: src/irsForms/Form.ts [hunk 1]
```diff
@@ -6,7 +6,9 @@ export type FormTag =
   'f1040sb' |
   'f1040sd' |
   'f1040se' |
-  'f1040sei'
+  'f1040sei' |
+  'f1040s8'
```
Comment (zakpatterson) [inline src/irsForms/Form.ts:10]: I was about to mention this should be the full number, but you've accurately captured the filename as it's downloaded from IRS, so this is good!

File: src/irsForms/Main.ts [hunk 2]
```diff
@@ -44,6 +46,14 @@ export const getSchedules = (f1040: F1040, state: Information): Form[] => {
     attachments = [...attachments, eic]
   }
 
+  const childtaxcreditws = new ChildTaxCreditWorksheet(f1040)
+  const schedule8812 = new Schedule8812(state.taxPayer, f1040)
+  if (f1040.dependents.reduce((total, dependent) => (childtaxcreditws.qualifiesChild(dependent) || childtaxcreditws.qualifiesOther(dependent) ? total + 1 : total), 0) > 0) {
```
Comment (zakpatterson) [inline src/irsForms/Main.ts:51]: line break

File: src/irsForms/Schedule8812.ts [hunk 3]
```diff
@@ -1,11 +1,101 @@
...
+  l3 = (): number | undefined => computeField(this.l1()) - computeField(this.l2()) > 0 ? computeField(this.l1()) - computeField(this.l2()) : undefined
+
+  l4 = (): number | undefined => computeField(this.f1040.childTaxCreditWorksheet?.numberQualifyingChildren()) > 0 ? computeField(this.f1040.childTaxCreditWorksheet?.numberQualifyingChildren()) * 1400 : undefined
+
+  l5 = (): number | undefined => Math.min(computeField(this.l3()), computeField(this.l4()))
+
+  // This is a horrible, horrible line
+  // have net earnings from self-employment and used optional methods => report pub972 Earned Income Worksheet (even if taking EIC)
+  // Taking EIC => completed worksheet B? => worksheet B line 4b, plus combat pay, minus clergy housing and meals, else EIC step 5 plus combat pay
+  // Not taking EIC? => pub 972 Earned Income Worksheet if self employed, or filing Schedule SE, or filing Schedule C as a "statutory employee" , else Form 1040 l1, minus fellowships, income as an inmate, deferred compensation, and Medicaid waiver, unless you chose to inclue the Medicaid waivers, and add combat pay in as well
```
Comment (zakpatterson) [inline src/irsForms/Schedule8812.ts:32]: Is this comment still needed? line doesn't seem too bad

File: src/irsForms/worksheets/ChildTaxCreditWorksheet.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+export default class ChildTaxCreditWorksheet {
+  f1040: F1040
+
+  constructor (f1040: F1040) {
+    this.f1040 = f1040
+  }
+
+  qualifiesChild = (d: Dependent): boolean =>
+    d.qualifyingInfo !== undefined && (CURRENT_YEAR - d.qualifyingInfo.birthYear < QualifyingDependents.childMaxAge)
```
Comment (zakpatterson) [inline src/irsForms/worksheets/ChildTaxCreditWorksheet.ts:14]: This is good. I thought it should be <= reading this line, but the instructions say under 19 at the end of 2020, so this accurately excludes a child who is 19. Now that I'm seeing it again, `maxAge` sounds inclusive to me. But a child with age 19 is older than 19 years and therefore should be excluded. Kind of depends on the reader. Maybe just a comment where maxAge is defined would help.

File: src/irsForms/Schedule8812.ts [hunk 5]
```diff
@@ -1,11 +1,101 @@
...
+
+  l2 = (): number | undefined => computeField(this.f1040.l19())
+
+  l3 = (): number | undefined => computeField(this.l1()) - computeField(this.l2()) > 0 ? computeField(this.l1()) - computeField(this.l2()) : undefined
+
+  l4 = (): number | undefined => computeField(this.f1040.childTaxCreditWorksheet?.numberQualifyingChildren()) > 0 ? computeField(this.f1040.childTaxCreditWorksheet?.numberQualifyingChildren()) * 1400 : undefined
+
+  l5 = (): number | undefined => Math.min(computeField(this.l3()), computeField(this.l4()))
+
+  // This is a horrible, horrible line
```
Comment (zakpatterson) [inline src/irsForms/Schedule8812.ts:29]: Is this comment still needed? Seems ok?

File: src/irsForms/Form.ts (hunk 1 above)
Comment (NathanielLovin) [reply src/irsForms/Form.ts:10]: Yep, I just copied the IRS filename.

**Comment by zakpatterson** [comment]
Hi, I wanted to be cautious of requesting your time, so I went ahead and made some suggested changes here:

https://github.com/NathanielLovin/UsTaxes/pull/2

After a while I'll merge those commits in here, or if you would like you're free to merge them in your fork as well which will bring them in here. 

Thanks!

**Review by zakpatterson — APPROVED** [review]

**Review by zakpatterson — APPROVED** [review]

**Review by thegrims — APPROVED** [review]

**Pull request merged**